### PR TITLE
[SPARK-25114][Core] Fix RecordBinaryComparator when subtraction between two words is divisible by Integer.MAX_VALUE.

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

https://github.com/apache/spark/pull/22079#discussion_r209705612 It is possible for two objects to be unequal and yet we consider them as equal with this code, if the long values are separated by Int.MaxValue.
This PR fixes the issue.

## How was this patch tested?
Add new test cases in `RecordBinaryComparatorSuite`.

### CODE REVIEW

**Comment by jiangxb1987** [comment]
cc @mridulm @squito

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java
Comment (kiszk) [inline sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java]: How about the following change to minimize and localize the change?
```
long res = Platform.getLong(leftObj, leftOff + i) -
             Platform.getLong(rightObj, rightOff + i);
if (res != 0) return res > 0 ? 1 : -1;
```

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java
Comment (kiszk) [inline sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java]: How about the following change to minimize and localize the change?
```
int res = (Platform.getByte(leftObj, leftOff + i) & 0xff) -
             (Platform.getByte(rightObj, rightOff + i) & 0xff);
if (res != 0) return res;
```

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java
Comment (kiszk) [inline sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java]: ditto

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java
Comment (mridulm) [reply sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java]: We can restrict scope of 'res' as an 'int' : and avoid the type promotions/conversions.

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java
Comment (mridulm) [reply sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java]: The subtraction is buggy due to potential overflow.
Why not simply use:
```
  final long v1 = Platform.getLong(leftObj, leftOff + i);
  final long v2 = Platform.getLong(rightObj, rightOff + i);
  if (v1 != v2) {
    return v1 > v2 ? -1 : 1;
  }
```

Comment (kiszk) [reply sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java]: +1 for no subtraction.

**Comment by jiangxb1987** [comment]
@squito I've created a new JIRA task and updated the title, thanks for reminding!

**Comment by kiszk** [comment]
LGTM

**Comment by bersprockets** [comment]
Should there be a test, or do other sorting-related tests cover this indirectly?

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java
Comment (gatorsmile) [inline sql/catalyst/src/main/java/org/apache/spark/sql/execution/RecordBinaryComparator.java]: Let us add a test suite for this function as you mentioned in the line 25. : )

**Review by srowen — APPROVED** [review]

**Comment by mridulm** [comment]
LGTM pending Xiao Li's excellent suggestion :-)

**Comment by jiangxb1987** [comment]
ping @gatorsmile @mridulm @squito

**Comment by squito** [comment]
the added tests are good.  This is pretty nit-picky, but looking at the whole test suite, are there any tests that check for anything other than the first byte (or array length)?  Seems the longer cases -- MultipleColumnRow, MixedColumns, ArrayColumn -- fail in a very early check.  NullColumn at least checks the last column.  kind of tedious test cases, but sometimes important in these low-level things as off-by-one errors are so easy.  (or maybe I just don't properly understand these test cases ...)

anyway I don't think that needs to hold this up more, just some thoughts.  lgtm, thanks for working on this.

**Comment by jiangxb1987** [comment]
Thanks @squito I've added another test case to cover when the last byte differs.

**Comment by jiangxb1987** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merged to master and 2.3

**Pull request closed without merging**
